70_FLIPFluids/app_resources/gridUtils.hlsl

```diff
@@ -0,0 +1,67 @@
+#ifndef _FLIP_EXAMPLE_GRID_UTILS_HLSL
+#define _FLIP_EXAMPLE_GRID_UTILS_HLSL
+
+#ifdef __HLSL_VERSION
+struct SGridData
+{
+    float gridCellSize;
+    float gridInvCellSize;
+    float pad0[2];
+
+    int4 particleInitMin;
+    int4 particleInitMax;
+    int4 particleInitSize;
+
+    float4 worldMin;
+    float4 worldMax;
+    int4 gridSize;
+};
+
+static const float POSITION_EPSILON = 1e-4;
+
+float4 clampPosition(float4 position, float4 gridMin, float4 gridMax)
+{
+    return float4(clamp(position.xyz, gridMin.xyz + POSITION_EPSILON, gridMax.xyz - POSITION_EPSILON), 1);
+}
+
+inline uint cellIdxToFlatIdx(int3 index, int4 gridSize)
+{
+    uint3 idxClamp = clamp(index, (int3)0, gridSize.xyz - 1);
+    return idxClamp.x + idxClamp.y * gridSize.x + idxClamp.z * gridSize.x * gridSize.y;
+}
+
+inline int3 flatIdxToCellIdx(uint id, int4 gridSize)
+{
+    int x = id % gridSize.x;
+    int y = id / gridSize.x % gridSize.y;
+    int z = id / (gridSize.x * gridSize.y);
+    return int3(x, y, z);
+}
+
+inline float3 cellIdxToWorldPos(int3 index, SGridData data)
+{
+    return data.worldMin.xyz + (index + 0.5f) * data.gridCellSize;
```

Comment:
we tend not to like implicit conversions between float and int

therefore its good to write `(float32_t3)index`, `float32_t3(index)` or `nbl::hlsl::_static_cast<float32_t3>(index)` to show you intend that to happen